### DISCUSSION TITLE
Example for reading a parameterized query / NO Spring app

### DISCUSSION THREAD

**Discussion by rheydenr** [graphql-java/graphql-java#2696] [General]
Hi all, I'm new to GraphQL and tried out some first exercises with this lib. But despite reading through documentation and examples I didn't get the following problem solved: I want to get a token from a remote server with the following query:

`query {
    token (
        username: "myuser",
        password: "mypwd"
    )
}
`
If I send this query with a web tool I get a valid answer (so, the server is working and the query is valid). But I couldn't get the mechanics how to do this in Java, esp. with the two parameters inside the query. Is there any Java solution example *without* Spring (plain Java with GraphQL lib)? Any hints are welcome. Thanks in advance.

Ralf.

**Comment by shuston** [comment]
Hi Ralf,

I’m not in front of code but I have used the American Express “nodes” package for client work. I don’t remember any Spring dependency but check it out. If you can use Spring I highly recommend Spring Graphql.

-Steve

On Jan 28, 2022, at 7:43 PM, Ralf Heydenreich ***@***.***> wrote:

﻿

Hi all, I'm new to GraphQL and tried out some first exercises with this lib. But despite reading through documentation and examples I didn't get the following problem solved: I want to get a token from a remote server with the following query:

query { token ( username: "myuser", password: "mypwd" ) }
If I send this query with a web tool I get a valid answer (so, the server is working and the query is valid). But I couldn't get the mechanics how to do this in Java, esp. with the two parameters inside the query. Is there any Java solution example without Spring (plain Java with GraphQL lib)? Any hints are welcome. Thanks in advance.

Ralf.

**Reply by rheydenr** [reply]
Hi Steve, thanks for your reply. Spring isn't an option since I'm in an OSGi environment. So, I'll look into the American Express lib. Seems to be quite interesting.
Cheers,
Ralf

**Comment by bbakerman** [comment]
So graphql-java is a server engine.  It's not a graphql client per se.

You can just send in a Http+JSON using any Java http client you like (UrlConnection / Apache Http Client ....

```
{
   "query:: "query { token ( username:\"myuser\", password: \"mypwd\" ) } "
}
```

and you will get back the `data` and or `errors as a JSON reply.

As the person above said, `nodes` is a graphql java client

There are more listed here : https://graphql.org/code/#java-kotlin

**Reply by rheydenr** [reply]
Hi, this could be a good approach which would be sufficient for me. But, if I try this with a simple HTTP client request, I get 

`
{"errors":[{"message":"Syntax Error: Unexpected <EOF>","extensions":{"category":"graphql"},"locations":[{"line":1,"column":1}]}]}
`

My data striing for the POST request is

`  
 String data = "{\"query\":: \"query {token(username: \"myuser\", password: \"mypwd\") } \"}";
`

The post itself is done with 

```      
HttpRequest request = HttpRequest.newBuilder() 
                .POST(HttpRequest.BodyPublishers.ofString(data))
                .uri(URI.create("http://localhost/example/graphql"))
                .setHeader("User-Agent", "Java 11 HttpClient Bot") 
                .header("Content-Type", "application/x-www-form-urlencoded")
                .build();
```

Did I miss some points?

Ralf.

**Reply by bbakerman** [reply]
> Syntax Error: Unexpected <EOF>

Means the query text is invalid - eg missing a closing `}` say .  It might be that the encoding of the query string is wrong.

You have used .header("Content-Type", "application/x-www-form-urlencoded") which it seems your server will accept (typically its `Content-Type :  application/json` but maybe your server is happy with form data.

I cant see how you created the `HttpRequest.BodyPublishers.ofString(data)` bit. 

you are on the right track however - the fact that you got back an error generated by the graphql server means that some of it is getting through into the graphql engine.

**Reply by bbakerman** [reply]
If you can, you need to get a dump of what you are truly sending over the wire in HTTP terms

And a graphql-java discussion place is not the best place for debugging HTTP clients - but we can help I guess

**Comment by rheydenr** [comment]
Finally, I got it :-)
Correct solution is:

```
        String data = "{\"query\":\"query { token(username: \\\"myuser\\\", password:\\\"mypassword\\\") }\"}";
        HttpRequest request = HttpRequest.newBuilder()
                .POST(HttpRequest.BodyPublishers.ofString(data))
                .uri(URI.create("http://localhost/OXID630/source/graphql"))
                .setHeader("User-Agent", "Java 11 HttpClient Bot") // add request header
                .header("Content-Type", "application/json")
                .build();

```
The problem was the wrong escaping of backslashes. Now it works.

Ralf.
